pkg/trimaran/lowriskovercommitment/lowriskovercommitment.go

```diff
@@ -0,0 +1,258 @@
+	return nil
+}
+
+// Checks and returns true if the pod is assigned to a node
+func isAssigned(pod *v1.Pod) bool {
+	return len(pod.Spec.NodeName) != 0
+}
+
+// computeRank : rank function for the LowRiskOverCommitment
+func (pl *LowRiskOverCommitment) computeRank(metrics []watcher.Metric, node *v1.Node, pod *v1.Pod) float64 {
+	// get all pods on node
+	nodeName := node.GetName()
+	podsOnNode := make([]*v1.Pod, 0)
+	pl.eventHandler.RLock()
+	for _, info := range pl.eventHandler.ScheduledPodsCache[node.GetName()] {
+		podsOnNode = append(podsOnNode, info.Pod)
+	}
+	pl.eventHandler.RUnlock()
+	// calculate risk based on requests and limits
+	nodeRequestsAndLimits := trimaran.GetNodeRequestsAndLimits(podsOnNode, node, pod)
+	riskCPU := pl.computeRisk(metrics, v1.ResourceCPU, watcher.CPU, node, nodeRequestsAndLimits)
+	riskMemory := pl.computeRisk(metrics, v1.ResourceMemory, watcher.Memory, node, nodeRequestsAndLimits)
+	rank := 1 - math.Max(riskCPU, riskMemory)
+
+	klog.V(6).InfoS("Node rank", "nodeName", nodeName, "riskCPU", riskCPU, "riskMemory", riskMemory, "rank", rank)
+
+	return rank
+}
+
+// computeRisk : calculate the risk of scheduling on node for a given resource
+func (pl *LowRiskOverCommitment) computeRisk(metrics []watcher.Metric, resourceName v1.ResourceName,
+	resourceType string, node *v1.Node, nodeRequestsAndLimits *trimaran.NodeRequestsAndLimits) float64 {
+
+	var riskLimit, riskLoad, totalRisk float64
+
+	defer func() {
+		klog.V(6).InfoS("Calculated risk", "node", klog.KObj(node), "resource", resourceName,
+			"riskLimit", riskLimit, "riskLoad", riskLoad, "totalRisk", totalRisk)
+	}()
+
+	nodeRequest := nodeRequestsAndLimits.NodeRequest
+	nodeLimit := nodeRequestsAndLimits.NodeLimit
+	nodeRequestMinusPod := nodeRequestsAndLimits.NodeRequestMinusPod
+	nodeLimitMinusPod := nodeRequestsAndLimits.NodeLimitMinusPod
+	nodeCapacity := nodeRequestsAndLimits.Nodecapacity
+
+	var request, limit, capacity, requestMinusPod, limitMinusPod int64
+	if resourceName == v1.ResourceCPU {
+		request = nodeRequest.MilliCPU
+		limit = nodeLimit.MilliCPU
+		requestMinusPod = nodeRequestMinusPod.MilliCPU
+		limitMinusPod = nodeLimitMinusPod.MilliCPU
+		capacity = nodeCapacity.MilliCPU
+	} else if resourceName == v1.ResourceMemory {
+		request = nodeRequest.Memory
+		limit = nodeLimit.Memory
+		requestMinusPod = nodeRequestMinusPod.Memory
+		limitMinusPod = nodeLimitMinusPod.Memory
+		capacity = nodeCapacity.Memory
+	} else {
+		// invalid resource
+		klog.V(6).InfoS("Unexpected resource", "resourceName", resourceName)
+		return 0
+	}
+
+	// (1) riskLimit : calculate overcommit potential load
+	excess := limit - request
+	if excess < 0 {
+		excess = 0
+	}
+	maxLoad := capacity
+	if limit < capacity {
+		maxLoad = limit
+	}
+	allowed := maxLoad - request
+	if allowed < 0 {
+		allowed = 0
+	}
+	if excess > 0 && allowed < excess {
+		riskLimit = 1 - float64(allowed)/float64(excess)
```

Comment:
Is that the higher riskLimit is, the more risk it has? If so, if a pod request 0.001 cpu, and limits a whole node's allocatable cpu, the riskLimit is ~0 in this case?

```
	// node's capacity: 1
	// pod's req/limits: 0.01/1
	// an existing pod with req/limits: 0.01/0.01
	// limit = 1.01 (1 + 0.01)
	// request = 0.01
	// excess = 1
	// maxLoad = 1
	// allowed = 0.99
```

Reply:
Yes, the risk is almost zero in this case. And, that makes sense. The reasoning is that if the pod is placed on that node it should be able to burst from its allocated 0.01 to the full node capacity without any risk.

Reply:
Hold on here... why the pod can burst to the full node capacity is at a low risk?

If a pod is specifying a req/lim request as 0.001/1, it implies it's somehow cheating the system's quota system, and if we pack pods to this kind to the same node, they may co-burst at the same time and that yields OutOfMemory or(and) CPU throttling.